sys/include/net/lwm2m.h

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (C) 2018 Beduino Master Projekt - University of Bremen
+ *               2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup     net_lwm2m LWM2M
+ * @ingroup      net
+ * @brief        Lightweight Machine to Machnine (LWM2M) definitions
+ * @{
+ *
+ * @file
+ * @brief       LWM2M header definitions
+ *
+ * @author      Christian Manal <manal@uni-bremen.de>
+ * @author      Leandro Lanzieri <leandro.lanzieri@haw-hamburg.de>
+ */
+
+
+#ifndef NET_LWM2M_H
+#define NET_LWM2M_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup lwm2m_conf     LWM2M compile configurations
+ * @ingroup config
+ * @{
+ */
+
+/**
+ * @brief CoAP default port of the LWM2M server
+ */
+#ifndef LWM2M_STANDARD_PORT
+#define LWM2M_STANDARD_PORT "5683"
+#endif
+
+/**
+ * @brief CoAPS default port of the LWM2M server
+ */
+#ifndef LWM2M_DTLS_PORT
+#define LWM2M_DTLS_PORT     "5684"
+#endif
+
+/**
+ * @brief CoAP default port of the LWM2M bootstrap server
+ */
+#ifndef LWM2M_BSSERVER_PORT
+#define LWM2M_BSSERVER_PORT "5685"
+#endif
+
+/**
+ * @brief Default port for the local LWM2M instance
+ */
+#ifndef LWM2M_LOCAL_PORT
+#define LWM2M_LOCAL_PORT    "5683"
+#endif
+
+/**
+ * @brief Device name used to register at the LWM2M server
+ */
+#ifndef LWM2M_DEVICE_NAME
+#define LWM2M_DEVICE_NAME "testRIOTDevice"
+#endif
+
+/**
+ * @brief Lifetime of the device object on the LWM2M server
+ */
+#ifndef LWM2M_DEVICE_TTL
+#define LWM2M_DEVICE_TTL 300
+#endif
+
+/**
+ * @brief LwM2M server URI to register/bootstrap with
+ *
+ * @note The host part of the URI MUST be a valid IPv6 address. Host names can
+ * not be resolved at this time.
+ */
+#ifndef LWM2M_SERVER_URI
+#define LWM2M_SERVER_URI "coap://[fd00:dead:beef::1]"
+#endif
+
+/**
+ * @brief Numeric ID of LWM2M_SERVER_URI
+ */
+#ifndef LWM2M_SERVER_ID
+#define LWM2M_SERVER_ID 10
+#endif
+
+/**
+ * @brief Alternate path to place LWM2M resources
+ */
+#ifndef LWM2M_ALT_PATH
+#define LWM2M_ALT_PATH  NULL
+#endif
+
+/**
+ * @brief Specifies if the server at @ref LWM2M_SERVER_URI is a bootstrap server
+ * or not.
+ *
+ * Valid values:
+ * - 0: Server is not bootstrap
+ * - 1: Server is bootstrap
+ */
+#ifndef LWM2M_SERVER_IS_BOOTSTRAP
+#define LWM2M_SERVER_IS_BOOTSTRAP 0
```

Comment:
Is this macro really necessary? Why not just use LWM2M_BOOTSTRAP?

If it is necessary, shouldn't the logical value of this macro be the same as LWM2M_BOOTSTRAP? Presently, LWM2M_BOOTSTRAP is defined in the example `Makefile`, which conflicts with this definition.

Also, LWM2M_BOOTSTRAP is used for conditionally including code. If I don't define it, the resulting image is smaller by 2096 bytes on samr21-xpro, which is significant.

Reply:
Good point. I agree we need to change this, and only have one input for this configuration.

Wakaama checks if the macro is defined, so if we choose to only use `LWM2M_BOOTSTRAP`, then we shouldn't give it a default value.
Maybe something like:
```C
/**
 * @brief Define to 1 to specify that LWM2M_SERVER_URI is a bootstrap server
 */
#ifdef DOXYGEN
#define LWM2M_BOOTSTRAP
#endif
```
Then every application can set this.

Another option could be to keep a macro that is not directly read by Wakaama (like `LWM2M_SERVER_BOOTSTRAP`) and translate it to `LWM2M_BOOTSTRAP`:

```C
#if LWM2M_SERVER_BOOTSTRAP == 1
#define LWM2M_BOOTSTRAP    1
#endif
```

Reply:
I would avoid the RIOT-specific definition, and use LWM2M_BOOTSTRAP directly. See the definition of DEBUG_ASSERT_VERBOSE in [assert.h](https://github.com/RIOT-OS/RIOT/blob/master/core/include/assert.h) for similar use of `#ifdef DOXYGEN`. I like the documentation and suggestion for how to use it.